registry/storage/driver/storj/storj.go

```diff
@@ -0,0 +1,589 @@
+					DriverName: driverName,
+					Path:       sourcePath,
+				}
+			} else if strings.Contains(err.Error(), "object already exists") { // TODO have this error in uplink
+				_, err := project.DeleteObject(ctx, d.bucket, storjKey(destPath))
+				if err != nil {
+					return err
+				}
+				continue
+			}
+			return err
+		}
+		return nil
+	}
+}
+
+// Delete recursively deletes all objects stored at "path" and its subpaths.
+func (d *driver) Delete(ctx context.Context, path string) error {
+	project, err := d.openProject(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() { err = errs.Combine(err, project.Close()) }()
+
+	iterator := project.ListObjects(ctx, d.bucket, &uplink.ListObjectsOptions{
+		Prefix:    storjKey(path) + "/",
+		Recursive: true,
+	})
+
+	found := false
+	for iterator.Next() {
+		found = true
+		item := iterator.Item()
+		_, err := project.DeleteObject(ctx, d.bucket, item.Key)
+		if err != nil {
+			return err
+		}
+	}
+	if err := iterator.Err(); err != nil {
+		return err
+	}
+
+	if found {
+		return nil
+	}
+
+	object, err := project.DeleteObject(ctx, d.bucket, storjKey(path))
+	if err != nil {
+		return err
+	}
+
+	if object == nil {
+		return storagedriver.PathNotFoundError{
+			DriverName: driverName,
+			Path:       path,
+		}
+	}
+	return nil
+}
+
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	// TODO most probably we can make linksharing link
+	return "", storagedriver.ErrUnsupportedMethod{}
+}
+
+// Walk traverses a filesystem defined within driver, starting
+// from the given path, calling f on each file
+func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn) error {
+	project, err := d.openProject(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() { err = errs.Combine(err, project.Close()) }()
+
+	prefix := storjKey(from)
+
+	if prefix != "" {
+		prefix += "/"
+	}
+
+	return d.doWalk(ctx, project, prefix, f)
+}
+
+func (d *driver) doWalk(ctx context.Context, project *uplink.Project, prefix string, f storagedriver.WalkFn) error {
+	storjPrefix := storjKey(prefix)
+
+	// TODO could we do this with single recursive request?
+	objects := project.ListObjects(ctx, d.bucket, &uplink.ListObjectsOptions{
+		Prefix: storjPrefix,
+	})
+
+	for objects.Next() {
+		item := objects.Item()
+
+		path := "/" + strings.TrimRight(item.Key, "/")
+		fileInfo := storagedriver.FileInfoInternal{FileInfoFields: storagedriver.FileInfoFields{
+			Path:    path,
+			Size:    item.System.ContentLength,
+			ModTime: item.System.Created,
+			IsDir:   item.IsPrefix,
+		}}
+		err := f(fileInfo)
+		if err != nil {
+			if err == storagedriver.ErrSkipDir {
+				continue
+			}
+			return err
+		}
+
+		if item.IsPrefix {
+			err = d.doWalk(ctx, project, item.Key, f)
```

Comment:
I believe this should be `d.doWalk(ctx, project, path, f)`. It would explain why we're able to walk the current directory structure, but not recursively as needed by the `_catalog` endpoint.

Reply:
hmmm.... nvm. key and path should be the same